### PR TITLE
perf(eventstore): get_events fetches event data from nodestore

### PR DESCRIPTION
**Pull request by lynnagara**
These queries yield 10+ columns from Snuba (including all
of the tag data) and are frequently taking a long time.
Instead we now fetch the minimal column set from Snuba, then use
those IDs to populate the event bodies from nodestore. Users of
eventstore.get_events() should no longer manually bind nodes since this
has already been done.

This depends on improvements to the message / search_message field https://github.com/getsentry/sentry/pull/16473

### CODE REVIEW

File: tests/sentry/eventstore/snuba_discover/test_backend.py [hunk 1]
```diff
@@ -74,45 +73,6 @@ def setUp(self):
 
         self.eventstore = SnubaDiscoverEventStorage()
 
-    def test_get_events(self):
```
Comment (lynnagara) [inline tests/sentry/eventstore/snuba_discover/test_backend.py]: These were unnecessary duplicates of snuba/test_backend.py

File: tests/snuba/models/test_event.py [hunk 2]
```diff
@@ -78,26 +77,3 @@ def test_bind_nodes(self):
         event.bind_node_data()
         assert event.data._node_data is not None
         assert event.data["user"]["id"] == u"user1"
-
-    def test_event_with_no_body(self):
```
Comment (lynnagara) [inline tests/snuba/models/test_event.py]: Since we don't have sampling as a feature anymore, I don't think this is a valid use case, and much of our product already depends on the events being present in nodestore and would otherwise be breaking if the event bodies were missing.

File: src/sentry/models/event.py [hunk 3]
```diff
@@ -496,7 +493,9 @@ def datetime(self):
     def message(self):
         if "message" in self.snuba_data:
             return self.snuba_data["message"]
-        return self.data.get("message")
+        from sentry.event_manager import EventManager
+        event_manager = EventManager(self.data.data)
```
Comment (fpacifici) [inline src/sentry/models/event.py]: Isn't EventManager depending on Event itself ? Could we avoid this dependency by moving get_search_message ? search_message feels like a property of an event.

File: src/sentry/eventstore/snuba/backend.py [hunk 4]
```diff
@@ -62,7 +62,9 @@ def get_events(
         )
 
         if "error" not in result:
-            return [SnubaEvent(evt) for evt in result["data"]]
+            events = [SnubaEvent(evt) for evt in result["data"]]
```
Comment (fpacifici) [inline src/sentry/eventstore/snuba/backend.py]: Could you elaborate on how you noticed the performance issue? Knowing that the timeout is around 30 seconds, I would be surprised if the culprit was fetching multiple columns instead and I would be less surprised if the issue was the select process, which means the queries are still likely to timeout. Could this be related to the max_thread parameter?

File: src/sentry/eventstore/snuba/backend.py [hunk 5]
```diff
@@ -62,7 +62,9 @@ def get_events(
...
+            self.bind_nodes(events)
```
Comment (fpacifici) [inline src/sentry/eventstore/snuba/backend.py]: This way we would be doing one request to bigtable per event. I guess, it would be more efficient to rely on nodestore.get_multi that makes one query and streams the results.
Also would the performance improvement be still relevant if we fetch like 1000 events (which means blasting bigtable with 1000 requests)

**Review by fpacifici — CHANGES_REQUESTED** [review]

File: src/sentry/eventstore/snuba/backend.py (hunk 4 above)
Comment (lynnagara) [reply src/sentry/eventstore/snuba/backend.py]: I think this is more about increasing the speed of existing queries, than reducing the count of timed out queries (hopefully we don't have too many of those). I've seen a number of transactions that look similar to this this https://sentry.io/organizations/sentry/eventsv2/snuba:7c55cef95fef4fafb2444c27376596fe. The 2nd query is significantly slower than the first since it's getting all columns https://sentry.io/organizations/sentry/eventsv2/snuba:7d55be78a00d43329937c42fcba9d2d5. I've changed the get_events implementation to only ever get the minimal columns from Snuba and the rest from Nodestore.

File: src/sentry/eventstore/snuba/backend.py (hunk 5 above)
Comment (lynnagara) [reply src/sentry/eventstore/snuba/backend.py]: There were 2 place where we were previously fetching more than 100 events:
- unmerge: This fetches 500 events at a time, since it was already doing bind_nodes immediately after the eventstore fetch (which has been removed) there should be no difference in performance
- nodestore deletions: Rewrote to not use eventstore since we only need to get 2 columns

Comment (fpacifici) [reply src/sentry/eventstore/snuba/backend.py]: If we are doing a multi_get (which we are) I think we should be fine with that. But we should keep an eye on this.

File: src/sentry/api/endpoints/group_hashes.py [hunk 6]
```diff
@@ -79,14 +79,17 @@ def __handle_results(self, project_id, group_id, user, results):
...
-        }
+        event = SnubaEvent(
+            {
+                "timestamp": result["latest_event_timestamp"],
+                "event_id": result["event_id"],
+                "group_id": group_id,
+                "project_id": project_id,
+            }
+        )
+        event.bind_node_data()
```
Comment (fpacifici) [inline src/sentry/api/endpoints/group_hashes.py]: What does this become after SnubaEvent is gone ?

File: src/sentry/deletions/defaults/group.py [hunk 7]
```diff
@@ -25,34 +26,33 @@ def chunk(self):
...
-            ),
-            limit=self.DEFAULT_CHUNK_SIZE,
+        events = raw_query(
+            selected_columns=["event_id", "timestamp"],
+            filter_keys={"project_id": [self.project_id], "group_id": [self.group_id]},
+            conditions=conditions,
             referrer="deletions.group",
             orderby=["-timestamp", "-event_id"],
-        )
+        )["data"]
```
Comment (fpacifici) [inline src/sentry/deletions/defaults/group.py]: I don't think we should bypass eventstore to fetch events. See my comment on PR https://github.com/getsentry/sentry/pull/16284.

Let's first see if loading everything from eventstore is a problem. Maybe we are fine with a multi_get. If it is a problem we should revisit the issue of loading a small number of fields on a big number of events through eventstore.

Comment (lynnagara) [reply src/sentry/deletions/defaults/group.py]: I'll revert this commit, and we can observe if it proves to be a problem before attempting something like this.

File: src/sentry/eventstore/snuba/backend.py (hunk 5 above)
Comment (lynnagara) [reply src/sentry/eventstore/snuba/backend.py]: The only place we are adding a multi get where where we weren't (at some later point outside eventstore) is in the group deletions task, we should keep an eye on that.

File: src/sentry/api/endpoints/group_hashes.py (hunk 6 above)
Comment (lynnagara) [reply src/sentry/api/endpoints/group_hashes.py]: This has been fixed in https://github.com/getsentry/sentry/pull/16286, will rebase to pull in those changes

**Review by fpacifici — CHANGES_REQUESTED** [review]
So after the discussion we had I believe you are going to still support binding nodes in some cases. RIght ?

File: src/sentry/eventstore/base.py [hunk 8]
```diff
@@ -117,6 +92,11 @@ def get_events(
         """
         raise NotImplementedError
 
+    def get_unfetched_events(
+        self, filter, orderby=None, limit=100, offset=0, referrer="eventstore.get_unfetched_events"
+    ):
+        raise NotImplementedError
```
Comment (fpacifici) [inline src/sentry/eventstore/base.py]: Docstring here is specifically important to explain the difference with get_events

File: src/sentry/eventstore/models.py [hunk 9]
```diff
@@ -197,6 +201,29 @@ def location(self):
             return self._snuba_data[column]
         return super(Event, self).location
 
+    def get_search_message(self, event_metadata=None, culprit=None):
```
Comment (fpacifici) [inline src/sentry/eventstore/models.py]: Is this change actually needed for fetching data from nodestore ?

File: src/sentry/eventstore/snuba/backend.py [hunk 10]
```diff
@@ -48,20 +48,59 @@ class SnubaEventStorage(EventStorage):
...
         orderby=None,
         limit=DEFAULT_LIMIT,
         offset=DEFAULT_OFFSET,
         referrer="eventstore.get_events",
     ):
         """
-        Get events from Snuba.
+        Get events from Snuba, with node data loaded.
         """
+        return self.get_events(
```
Comment (fpacifici) [inline src/sentry/eventstore/snuba/backend.py]: is this a call to __get_events instead ?

**Review by fpacifici — CHANGES_REQUESTED** [review]
Failing tests are related.

File: src/sentry/eventstore/snuba/backend.py (hunk 10 above)
Comment (lynnagara) [reply src/sentry/eventstore/snuba/backend.py]: uhhh, yes it should be

File: src/sentry/eventstore/models.py (hunk 9 above)
Comment (lynnagara) [reply src/sentry/eventstore/models.py]: It's a dependency. This was kind of a hack before since the message from snuba is different from the message from nodestore. We're no longer providing a way to explicitly say that you want the column to be loaded from snuba (a good thing since this was being abused before). I'm going to try to improve this a bit more in a separate PR.

Comment (lynnagara) [reply src/sentry/eventstore/models.py]: See https://github.com/getsentry/sentry/pull/16473 for a more proper fix for the event.message / event.search_message behaviour

**Review by fpacifici — APPROVED** [review]
